File: Android/app/src/main/java/org/gdgsp/other/Other.java
/*
 * Copyright (C) 2016 Alefe Souza <http://alefesouza.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gdgsp.other;

import android.app.Activity;
import android.app.PendingIntent;
import android.content.ActivityNotFoundException;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager.NameNotFoundException;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Build;
import android.preference.PreferenceManager;
import android.support.customtabs.CustomTabsIntent;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Toast;
import org.gdgsp.R;
import org.gdgsp.lib.*;

/**
 * Aqui ficam métodos e atributos utilizados em várias classes do aplicativo.
 */
public class Other extends Activity {
    /**
     * Cor primária do aplicativo, usada como cor da toolbar da Custom Tab.
     */
	public static final int colorPrimary = 0xff008bf8;

	/**
	 * String que tem que aparecer no final de todas as URLs no aplicativo.
	 */
	public static String getFinalUrl(Context context) {
		return "appversion=" + getAppVersion(context) + "&platform=android&sdkint=" + getOSVersion();
	}

    /**
     * Métodos que retorna a URL de receber os eventos.
     * @param context Contexto usado para receber uma string no string.xml
     * @return A URL final para receber os eventos.
     */
    public static String getEventsUrl(Context context) {
		return "http://" + context.getString(R.string.backend_url) + "api/events.php?meetupid=" + context.getString(R.string.meetup_id) + "&" + getFinalUrl(context);
    }

    /**
     * Método que retorna a URL para fazer RSVP.
     * @param context Contexto usado para receber uma string no string.xml
     * @param id O id do evento que o usuário está fazendo RSVP.
     * @return A URL final para fazer RSVP no evento que o usuário quer ir.
     */
	public static String getRSVPUrl(Context context, int id) {
        return "http://" + context.getString(R.string.backend_url) + "api/rsvp.php?meetupid=" + context.getString(R.string.meetup_id) + "&eventid=" + id + "&" + getFinalUrl(context);
    }

    /**
     * Método que retorna a URL das pessoas que vão para o evento.
     * @param context Contexto usado para receber uma string no string.xml
     * @param id O id do evento.
     * @return A URL final para receber as respostas das pessoas sobre o evento.
     */
    public static String getRSVPSUrl(Context context, int id) {
        return "http://" + context.getString(R.string.backend_url) + "api/people.php?meetupid=" + context.getString(R.string.meetup_id) + "&eventid=" + id + "&" + getFinalUrl(context);
    }

    /**
     * Método que retorna a URL para fazer login.
     */
	public static String getLoginUrl(Context context) {
		return "http://" + context.getString(R.string.backend_url) + "api/login.php?meetupid=" + context.getString(R.string.meetup_id);
	}

	/**
	 * Método que retorna a URL para o administrador enviar notificação.
	 */
	public static String getNotificationUrl(Context context) {
		return "http://" + context.getString(R.string.backend_url) + "notifications/send.php?meetupid=" + context.getString(R.string.meetup_id);
	}

    /**
     * Método que retorna o refresh token armazenado nas configurações.
     * @param context Contexto usando para receber a configuração.
     */
	public static String getRefreshToken(Context context) {
		SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);

		return preferences.getString("refresh_token", "");
	}

    /**
     * Método que abre um link em uma Custom Tab
     * @param context Contexto usado
     * @param url URL a ser aberta
     */
	public static void openSite(Context context, String url) {
		url = url.startsWith("http") ? url : "http://" + url;
		
		CustomTabsIntent.Builder intentBuilder = new CustomTabsIntent.Builder();
		intentBuilder.setToolbarColor(colorPrimary);
		intentBuilder.setShowTitle(true);

		String shareLabel = context.getString(R.string.share);
		Bitmap icon = BitmapFactory.decodeResource(context.getResources(), R.drawable.ic_share);

		Intent actionIntent = new Intent(context.getApplicationContext(), ShareBroadcastReceiver.class);

		PendingIntent pendingIntent = PendingIntent.getBroadcast(context, 0, actionIntent, 0);
		intentBuilder.setActionButton(icon, shareLabel, pendingIntent);

		CustomTabActivityHelper.openCustomTab((Activity)context, intentBuilder.build(), Uri.parse(url), new WebviewFallback());
	}

    /**
     * Tenta abrir o app do Meetup com uma url, se não der certo usa o método acima.
     * @param context Contexto usado.
     * @param url URL a ser aberta.
     */
	public static void openMeetupApp(Context context, String url) {
		try {
			Intent intent = new Intent(Intent.ACTION_VIEW);
			intent.setPackage("com.meetup");
			intent.setData(Uri.parse(url));
			context.startActivity(intent);
		} catch (ActivityNotFoundException e) {
			openSite(context, url);
		}
	}

    /**
     * Retorna o SDKINT do Android
     */
	public static int getOSVersion() {
		return Build.VERSION.SDK_INT;
	}

    /**
     * Retorna a versão do aplicativo
     */
	public static String getAppVersion(Context context) {
		try {
			PackageInfo pInfo = context.getPackageManager().getPackageInfo(context.getPackageName(), 0);
			return pInfo.versionName;
		} catch (NameNotFoundException e) {}
		return null;
	}

    /**
     * Exibe uma caixa de mensagem simples.
     * @param context Contexto usado.
     * @param title Título da mensagem.
     * @param description Descrição da mensagem.
     */
	public static void showMessage(Context context, String title, String description) {
		final AlertDialog alertDialog = new AlertDialog.Builder(context)
			.setTitle(title)
			.setMessage(description)
			.setPositiveButton("OK", null)
			.create();
			
		alertDialog.show();
	}

    /**
     * Exibe um toast.
     * @param context Contexto usado.
     * @param message Mensagem do toast.
     */
	public static void showToast(Context context, String message) {
		Toast.makeText(context, message, Toast.LENGTH_LONG).show();
	}

    /**
     * Método que executa ao pressionar um card da tela inicial.
     * @param context Contexto usado.
     * @param title Título do evento.
     * @param url URL do evento.
     */
	public static void cardDialog(final Context context, final String title, final String url) {
		final String items[] ={"Compartilhar", "Copiar link", "Abrir no navegador"};
        AlertDialog.Builder alertDialog = new AlertDialog.Builder(context);
        LayoutInflater inflater = (LayoutInflater)context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        View convertView = inflater.inflate(R.layout.list_pressed, null);
        alertDialog.setView(convertView);
        alertDialog.setTitle(title);

		alertDialog.setItems(items, new DialogInterface.OnClickListener() {
				public void onClick(DialogInterface dialog, int item) {
					switch(item) {
						case 0:
							Intent shareIntent = new Intent(Intent.ACTION_SEND);
							shareIntent.setType("text/plain");
							shareIntent.putExtra(Intent.EXTRA_TEXT, title + " " + url);

							Intent chooserIntent = Intent.createChooser(shareIntent, context.getString(R.string.share));
							chooserIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
							
							context.startActivity(chooserIntent);
							break;
						case 1:
                            if(Build.VERSION.SDK_INT <= 10) {
                                android.text.ClipboardManager clipboard = (android.text.ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
                                clipboard.setText(url);
                            } else {
                                ClipboardManager clipboardManager = (ClipboardManager)context.getSystemService(Context.CLIPBOARD_SERVICE);
                                ClipData clipData = ClipData.newPlainText(url, url);
                                clipboardManager.setPrimaryClip(clipData);
                            }
							showToast(context, context.getString(R.string.link_copyed));
							break;
						case 2:
							Intent intent = new Intent(Intent.ACTION_VIEW);
							intent.setData(Uri.parse(url));
							context.startActivity(intent);
							break;
					}
				}
			});

		alertDialog.create();
        alertDialog.show();
	}

    /**
     * Retorna se o aplicativo está sendo executado em um tablet
     */
	public static boolean isTablet(Context context) {
		return context.getResources().getBoolean(R.bool.isTablet);
	}

    /**
     * Método que retorna a um parâmetro de uma URL.
     * @param url URL para extrair o parâmetro.
     * @param want Parâmetro a ser extraído.
     * @return
     */
	public static String getQuery(String url, String want) {
		String[] a = url.split("\\?");
		String[] b = a[1].split("&");
		for(String c : b) {
			if(c.split("=")[0].equals(want)) {
				return c.split("=")[1];
			}
		}
		return "";
	}
}

File: Xamarin/GDG-SP/GDG_SP/Other/Other.cs
/*
 * Copyright (C) 2016 Alefe Souza <http://alefesouza.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using GDG_SP.Resx;
using Plugin.Settings;
using Plugin.Share;
using System.Collections.Generic;
using System.Diagnostics;
using System.Net.Http;
using Xamarin.Forms;

namespace GDG_SP.Other
{
    /// <summary>
    /// Aqui ficam métodos e atributos utilizados em várias classes do aplicativo.
    /// </summary>
    public class Other
    {
        /// <summary>
        /// String que tem que aparecer no final de todas as URLs no aplicativo.
        /// </summary>
		public static string GetFinalUrl()
        {
            string platform = Device.OnPlatform("ios", "android", "wp");

            if (Device.Idiom == TargetIdiom.Desktop)
            {
                platform = "windows";
            }
            else if (Device.Idiom == TargetIdiom.Tablet && Device.OS == TargetPlatform.Windows)
            {
                platform = "windows81";
            }

            return "?meetupid=" + AppResources.MeetupId + "&appversion=" + DependencyService.Get<IDependencies>().GetAppVersion() + "&systemversion=" + DependencyService.Get<IDependencies>().GetOSVersion() + "&platform=" + platform + "&via=xamarin";
        }

        /// <summary>
        /// Métodos que retorna a URL de receber os eventos.
        /// </summary>
        /// <returns>A URL final para receber os eventos.</returns>
        public static string GetEventsUrl()
        {
            return "http://" + AppResources.BackendUrl + "api/events.php" + GetFinalUrl();
        }

        /// <summary>
        /// Método que retorna a URL para fazer RSVP.
        /// </summary>
        /// <param name="id">O id do evento que o usuário está fazendo RSVP.</param>
        /// <returns>A URL final para fazer RSVP no evento que o usuário quer ir.</returns>
        public static string GetRSVPUrl(int id)
        {
            return "http://" + AppResources.BackendUrl + "api/rsvp.php" + GetFinalUrl() + "&eventid=" + id;
        }

        /// <summary>
        /// Método que retorna a URL das pessoas que vão para o evento.
        /// </summary>
        /// <param name="id">O id do evento.</param>
        /// <returns>A URL final para receber as respostas das pessoas sobre o evento.</returns>
        public static string GetRSVPSUrl(int id)
        {
			return "http://" + AppResources.BackendUrl + "api/people.php" + GetFinalUrl() + "&eventid=" + id;
        }

        /// <summary>
        /// Método que retorna a URL para fazer login.
        /// </summary>
        public static string GetLoginUrl()
		{
			return "http://" + AppResources.BackendUrl + "api/login.php" + GetFinalUrl();
        }

        /// <summary>
        /// Método que retorna a URL para o administrador enviar notificação.
        /// </summary>
        public static string GetNotificationUrl()
        {
            return "http://" + AppResources.BackendUrl + "notifications/send.php" + GetFinalUrl();
        }

        /// <summary>
        /// Método que retorna o refresh token armazenado nas configurações para fazer um post junto com uma URL.
        /// </summary>
        /// <returns>FormUrlEncodedContent configurado para post do refresh token no servidor.</returns>
        public static FormUrlEncodedContent GetRefreshToken(bool withChannel = false)
        {
            var postData = new List<KeyValuePair<string, string>>();
            postData.Add(new KeyValuePair<string, string>("refresh_token", GetSetting("refresh_token")));

            if ((Device.Idiom == TargetIdiom.Desktop || (Device.Idiom == TargetIdiom.Tablet && Device.OS == TargetPlatform.Windows)) && withChannel)
            {
                postData.Add(new KeyValuePair<string, string>("ChannelUri", GetSetting("ChannelUri")));
            }

            return new FormUrlEncodedContent(postData);
        }

        /// <summary>
        /// Exibir uma caixa de mensagem é algo comum no aplicativo.
        /// </summary>
        /// <param name="message">Mensagem a ser exibida.</param>
        public static async void ShowMessage(string message, Page page)
        {
            await page.DisplayAlert("", message, "OK");
        }

        /// <summary>
        /// Método que abre uma URL.
        /// </summary>
        /// <param name="url">URL que deve ser aberta.</param>
        /// <param name="page">Page que "herdará" a navegação.</param>
        /// <param name="inApp">Se a URL deve ser aberta no aplicativo ou no navegador externo.</param>
        /// <param name="title">Título da página a ser aberta.</param>
        public static void OpenSite(string url, Page page, bool inApp = false, string title = "")
        {
            url = url.StartsWith("http") ? url : "http://" + url;
            if (inApp)
            {
                page.Navigation.PushAsync(new WebViewPage(url) { Title = title });
            }
            else
            {
				CrossShare.Current.OpenBrowser(url, null);
            }
        }

        /// <summary>
        /// Método que compartilha um link com o plugin CrossShare
        /// </summary>
        /// <param name="title">Título a ser exibido na tela de compartilhamento.</param>
        /// <param name="link">Link a ser compartilhado.</param>
		public static async void ShareLink(string title, string link)
		{
			await CrossShare.Current.ShareLink(link, "", title);
		}
        
        /// <summary>
        /// Método que retorna a URL do back-end, usado no GDG_SP.WinPhone para criar as live tiles, não conseguir acessar o AppResources diretamente no outro projeto.
        /// </summary>
        /// <returns>A URL do back-end.</returns>
        public static string GetBackendUrl()
        {
            return AppResources.BackendUrl;
        }

        /// <summary>
        /// Mesma coisa do método acima, porém retorna o app id do One Signal.
        /// </summary>
        /// <returns>A app id do One Signal.</returns>
        public static string GetOneSignalAppId()
        {
            return AppResources.OneSignalAppId;
        }

        /// <summary>
        /// Mesma coisa do método acima, porém a cor principal.
        /// </summary>
        /// <returns>A cor principal.</returns>
        public static string GetColorPrimary()
        {
            return AppResources.ColorPrimary;
        }

        /// <summary>
        /// Adiciona configuração.
        /// </summary>
        /// <param name="key">Chave da configuração.</param>
        /// <param name="value">Valor da configuração.</param>
        public static void AddSetting(string key, string value)
        {
            CrossSettings.Current.AddOrUpdateValue<string>(key, value);
        }

        /// <summary>
        /// Retorna uma configuração armazenada.
        /// </summary>
        /// <param name="key">Chave da configuração.</param>
        /// <returns>Valor da configuração.</returns>
        public static string GetSetting(string key)
        {
            return CrossSettings.Current.GetValueOrDefault<string>(key, "");
        }
        
        /// <summary>
        /// Método que retorna a um parâmetro de uma URL.
        /// </summary>
        /// <param name="url">URL para extrair o parâmetro.</param>
        /// <param name="what">Parâmetro a ser extraído.</param>
        public static string GetQuery(string url, string what)
        {
            string[] parts = url.Split(new char[] { '?', '&' });
            foreach (string p in parts)
            {
                if (p.Contains(what))
                {
                    return p.Split('=')[1];
                }
            }

            return null;
        }

        /// <summary>
        /// Xamarin não muda a altura com HorizontalOptions="FillAndExpand" no iOS igual no Windows Phone, tive a ideia de fazer isso
        /// </summary>
        /// <param name="actualWidth">Largura atual da imagem oculpando a tela toda.</param>
        /// <param name="width">Largura da imagem retornada pelo JSON.</param>
        /// <param name="height">Altura da imagem retornada pelo JSON.</param>
        /// <returns>A altura que a imagem deve ficar para manter a proporção tendo a largura como tela/coluna inteira.</returns>
        public static double GetHeightImage(double actualWidth, double width, double height)
        {
            double ratio = width / height;
            double targetWidth = actualWidth;

            double targetHeight = targetWidth / ratio;

            return targetHeight;
        }

        public static string GetImage(string name)
        {
            switch(Device.OS)
            {
                case TargetPlatform.Android:
                    name = "ic_" + name.ToLower();
                    break;
                case TargetPlatform.Windows:
                    name = "Assets/Images/" + name;
                    break;
            }
            string directory = name + ".png";
            return directory;
        }
    }
}
